hie-plugin-api/Haskell/Ide/Engine/Cradle.hs

```diff
@@ -17,5 +35,218 @@ findLocalCradle fp = do
   -- Get the cabal directory from the cradle
   cradleConf <- BIOS.findCradle fp
   case cradleConf of
-    Just yaml -> BIOS.loadCradle yaml
-    Nothing -> BIOS.loadImplicitCradle fp
\ No newline at end of file
+    Just yaml -> fixCradle <$> BIOS.loadCradle yaml
+
+    Nothing   -> cabalHelperCradle fp
+
+-- | Check if the given Cradle is a stack cradle.
+-- This might be used to determine the GHC version to use on the project.
+-- If it is a stack-Cradle, we have to use `stack path --compile-exe`
+-- otherwise we may ask `ghc` directly what version it is.
+isStackCradle :: Cradle -> Bool
+isStackCradle = (`elem` ["stack", "Cabal-Helper-Stack"]) . BIOS.actionName . BIOS.cradleOptsProg
+
+-- | Finds a Cabal v2-project, Cabal v1-project or a Stack project
+-- relative to the given FilePath.
+-- Cabal v2-project and Stack have priority over Cabal v1-project.
+-- This entails that if a Cabal v1-project can be identified, it is
+-- first checked whether there are Stack projects or Cabal v2-projects
+-- before it is concluded that this is the project root.
+-- Cabal v2-projects and Stack projects are equally important.
+-- Due to the lack of user-input we have to guess which project it
+-- should rather be.
+-- This guessing has no guarantees and may change any-time.
+findCabalHelperEntryPoint :: FilePath -> IO (Maybe (Ex ProjLoc))
+findCabalHelperEntryPoint fp = do
+  projs <- concat <$> mapM findProjects subdirs
+  case filter (\p -> isCabalNewProject p || isStackProject p) projs of
+    (x:_) -> return $ Just x
+    []    -> case filter isCabalOldProject projs of
+      (x:_) -> return $ Just x
+      []    -> return Nothing
+  where
+    -- | Subdirectories of a given FilePath.
+    -- Directory closest to the FilePath `fp` is the head,
+    -- followed by one directory taken away.
+    subdirs :: [FilePath]
+    subdirs = reverse . map joinPath . tail . inits $ splitDirectories (takeDirectory fp)
+
+    isStackProject (Ex ProjLocStackYaml {}) = True
+    isStackProject _ = False
+
+    isCabalNewProject (Ex ProjLocV2Dir {}) = True
+    isCabalNewProject (Ex ProjLocV2File {}) = True
+    isCabalNewProject _ = False
+
+    isCabalOldProject (Ex ProjLocV1Dir {}) = True
+    isCabalOldProject (Ex ProjLocV1CabalFile {}) = True
+    isCabalOldProject _ = False
+
+-- | Given a FilePath, find the Cradle the FilePath belongs to.
+--
+-- TODO: document how and why this works.
```

Comment:
Yeah, i am trying to know where the project build is triggered with no luck :smile:

Reply:
The line `packages <- runQuery projectPackages env` initialises the project contexts, e.g. configures projects enough such that we know the different packages in a project.
It produces the output one can see from HieWrapper.hs. However, afaik, this does not actually compile anything, I think it is only the configure step.
Actual compilation of the different units, e.g. library, executable happens on cradle basis: `unitInfos_ <- mapM (\unit -> runQuery (unitInfo unit) env) units`
Which should not compile the unit but download all dependencies and so, such that we know the targets in that unit.

Reply:
Is possible that cabal-helper is running the [`cabal build --dry-run`](https://github.com/DanielG/cabal-helper/blob/447814db7ecda25afa13a7a699a72c5223649d98/lib/Distribution/Helper.hs#L494) shown in the logs?

Reply:
But hie is passing the `unitId` to `cabal-helper` and it can be the cause of errors with components with the same name, maybe we should prefix it with the package name (not sure what field would give us it), to avoid ambiguities

Reply:
> 
> Is possible that cabal-helper is running the cabal build --dry-run shown in the logs?

It is, but I dont know to be honest. I am just using the High-Level API-

> 
> But hie is passing the unitId to cabal-helper and it can be the cause of errors with components with the same name, maybe we should prefix it with the package name (not sure what field would give us it), to avoid ambiguities

We are simply passing Units that have been obtained from a Package. Unless I am overlooking something, I feel like this is rather a minor bug of c-h. 

Thank you for thorough analysis, investigative questions and constructive feedback!